homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,326 @@
+        await hass.services.async_call(DEVICE_TRACKER_DOMAIN,
+                                       DEVICE_TRACKER_SEE, data)
+        return json_response([])
+
+
+def supports_encryption():
+    """Test if we support encryption."""
+    try:
+        import nacl   # noqa pylint: disable=unused-import
+        return True
+    except OSError:
+        return False
+
+
+def device_for_webhook_id(hass, webhook_id):
+    """Return the device name for the webhook ID."""
+    for device_name, device in hass.data[DOMAIN].items():
+        if device.get(CONF_WEBHOOK_ID) == webhook_id:
+            return device
+    return None
+
+
+async def async_setup(hass, config):
+    """Set up the mobile app component."""
+    conf = config.get(DOMAIN)
+
+    store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+    app_config = await store.async_load()
+    if app_config is None:
+        app_config = {}
+
+    hass.data[DOMAIN] = app_config
+
+    for name, device in app_config.items():
+        if CONF_WEBHOOK_ID in device:
+            webhook.async_register(hass, DOMAIN, 'Mobile app',
+                                   device[CONF_WEBHOOK_ID], handle_webhook)
+
+    if conf is not None:
+        hass.async_create_task(hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': config_entries.SOURCE_IMPORT}))
+
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up an mobile app entry."""
+    hass.http.register_view(RegisterDeviceView())
+
+    return True
+
+
+class RegisterDeviceView(HomeAssistantView):
+    """A view that accepts device registration requests."""
+
+    url = '/api/mobile_app/register'
+    name = 'api:mobile_app:register'
+
+    def __init__(self):
+        """Initialize the view."""
+
+    async def post(self, request):
+        """Handle the POST request for device registration."""
+        try:
+            req_data = await request.json()
+        except ValueError:
+            return self.json_message("Invalid JSON", HTTP_BAD_REQUEST)
+
+        hass = request.app['hass']
+
+        try:
+            data = REGISTER_DEVICE_SCHEMA(req_data)
+        except vol.Invalid as ex:
+            return self.json_message(
+                vol.humanize.humanize_error(req_data, ex),
+                HTTP_BAD_REQUEST)
+
+        device_id = data[ATTR_DEVICE_ID]
+
+        if device_id in hass.data[DOMAIN]:
+            return self.json(hass.data[DOMAIN][device_id])
+
+        resp = {}
+
+        webhook_id = generate_secret()
+
+        data[CONF_WEBHOOK_ID] = webhook_id
+
+        resp[CONF_WEBHOOK_ID] = webhook_id
+
+        if data[ATTR_SUPPORTS_ENCRYPTION] is True and supports_encryption():
+            secret = generate_secret(16)
+
+            data[CONF_SECRET] = secret
+
+            resp[CONF_SECRET] = secret
+
+        hass.data[DOMAIN][device_id] = data
+
+        store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
```

Comment:
We should only keep one instance of the store, because it has a write lock. Initialize the store in `async_setup` and pass it to the constructor of the view.